12508: Feature add segment rows flush config

karthi07 opened the pull request:
Adding new config "realtime.segment.flush.threshold.segment.rows" for real-time table to configure the segment threshold independent of max partitions

**Behavior:**
- If flush rows > 0, create a new DefaultFlushThresholdUpdater with the given flush size.
- If flush segment rows > 0, create a new FixedFlushThresholdUpdater with the given flush size.
- If flush segment size > 0, create a new SegmentSizeBasedFlushThresholdUpdater if not already created. Create only 1 per table because we want to maintain tuning information for the table in the updater.

**Legacy behavior:**
- If flush rows = 0, use segment size based flush threshold.
- If none of the above are set, create a new DefaultFlushThresholdUpdater.
   - If flush size > 0, create a new DefaultFlushThresholdUpdater with given flush size.

**Issue[12508]:**
https://github.com/apache/pinot/issues/12508
 When realtime.segment.flush.threshold.rows is used, the flush threshold for the new CONSUMING segment is determined by both this value and the max partitions consumed by any server. This is not very straight forward, and rebalancing a table could cause new consuming segment size to change.

In order to tackle this problem, we may add a new config (proposing realtime.segment.flush.threshold.segment.rows) which works independent of partitions consumed.

Jackie-Jiang reviewed (commented):
Thanks for the contribution! I feel the current way (not introduced in this PR) of setting default value for threshold is very confusing. Let me try to simplify that and you may work on the updated code

Jackie-Jiang commented:
#12684 cleans up the current threshold handling

karthi07 commented:
@Jackie-Jiang Your suggestions were helpful, I updated the docs and renamed the FixedThreshold. let me know if the docs need additional changes. 

Should I merge #12684 into my branch and update the default threshold calc, or should I wait till it gets merged into master?

Jackie-Jiang commented:
@karthi07 Just merged #12684. Can you please rebase this PR on top of that?

Jackie-Jiang reviewed (approved):
It looks good in general. I applied a commit for some detailed handling and extra testing
